File: android/app/src/main/kotlin/com/movie/discovery/movie_discovery/network/NetworkChannelHandler.kt
package com.movie.discovery.movie_discovery.network

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class NetworkChannelHandler : MethodChannel.MethodCallHandler {
    private val apiClient = ApiClient()
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "request" -> handleRequest(call, result)
            else -> result.notImplemented()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun handleRequest(call: MethodCall, result: MethodChannel.Result) {
        val baseUrl = call.argument<String>("baseUrl")
        val path = call.argument<String>("path")

        if (baseUrl.isNullOrEmpty()) {
            result.error("INVALID_ARGUMENTS", "baseUrl is required and cannot be empty", null)
            return
        }

        if (path.isNullOrEmpty()) {
            result.error("INVALID_ARGUMENTS", "path is required and cannot be empty", null)
            return
        }

        val method = call.argument<String>("method") ?: "GET"
        val queryParams = call.argument<Map<String, Any>>("queryParams")
        val headers = call.argument<Map<String, String>>("headers")

        scope.launch {
            try {
                val response = apiClient.request(
                    baseUrl = baseUrl,
                    path = path,
                    method = method,
                    queryParams = queryParams,
                    headers = headers
                )
                result.success(response)
            } catch (e: Exception) {
                result.error(
                    "NETWORK_ERROR",
                    e.message ?: "Unknown error",
                    null
                )
            }
        }
    }
}

File: ios/Runner/Network/NetworkChannelHandler.swift
import Flutter
import Foundation

class NetworkChannelHandler: NSObject, FlutterPlugin {
    private let apiClient = ApiClient()

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.movie.discovery/network",
            binaryMessenger: registrar.messenger()
        )
        let instance = NetworkChannelHandler()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "request":
            handleRequest(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleRequest(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Arguments must be a map",
                details: nil
            ))
            return
        }

        guard let baseUrl = args["baseUrl"] as? String, !baseUrl.isEmpty else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "baseUrl is required and cannot be empty",
                details: nil
            ))
            return
        }

        guard let path = args["path"] as? String, !path.isEmpty else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "path is required and cannot be empty",
                details: nil
            ))
            return
        }

        let method = args["method"] as? String ?? "GET"
        let queryParams = args["queryParams"] as? [String: Any]
        let headers = args["headers"] as? [String: String]

        apiClient.request(
            baseUrl: baseUrl,
            path: path,
            method: method,
            queryParams: queryParams,
            headers: headers
        ) { response in
            result(response)
        }
    }
}
